lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -43,13 +43,50 @@ using namespace tethys;
 
 class tethys::ScienceSensorsSystemPrivate
 {
+  //////////////////////////////////
+  // Functions for data manipulation
+
   /// \brief Reads csv file and populate various data fields
   /// \param[in] _ecm Immutable reference to the ECM
   public: void ReadData(const ignition::gazebo::EntityComponentManager &_ecm);
 
   /// \brief Generate octree from spatial data, for searching
   public: void GenerateOctrees();
 
+  /// \brief Barycentric interpolation in 3D, given 4 points on any arbitrary
+  /// tetrahedra.
+  /// \param[in] _p Position within the tetrahedra to interpolate for
+  /// \param[in] _xyzs XYZ coordinates of 4 vertices of a tetrahedra
+  /// \param[in] _values Data values at the 4 vertices
+  /// \return Interpolated value, or quiet NaN if inputs invalid.
+  public: float BarycentricInterpolate(
+    const Eigen::Vector3f &_p,
+    const std::vector<pcl::PointXYZ> &_xyzs,
+    const std::vector<float> &_values);
+
+  /// \brief Barycentric interpolation in 2D, given 3 points on any arbitrary
```

Comment:
2D is a valid case if the robot is exactly on the layer of data, as is the case in the default robot location - it's lying at z = 0, which is exactly where the data is, given that we pulled a lat/long from the data set to initialize the world. Given how slow the nominal velocity is (1 m/s), if a robot is traveling for a while at such a depth, the 2D case saves some computation over the 3D. For performance, there can even be some threshold set for when to just do 2D instead of 3D, if the robot is substantially closer to one layer than another.

Reply:
This definitely needs to be kept. While testing, I realized that this is a mathematical case for T, not just a coincidental happening with the data.
Whenever T has 1 row of zeros, that degenerates to 2D (points are on a plane).
Whenever T has 2 rows of zeros, that degenerates to 1D (points are in a line).
When T is all zeros, then all points are at the exact same point.
0d0b14d generalizes the 2D case (previously it was just checking for the z row being zero) and handles all 3 degenerative cases.